### DISCUSSION TITLE
We should decide the fate of the 3D threaded integrator

### DISCUSSION THREAD

**Discussion by graeme-winter** [dials/dials#1723] [General]
Either
1. remove it completely or
2. move to using it as the default / only implementation

The second of these would be a very bold move. The first is nominally some sunk cost but I feel that the confusion of it being present but never used outweighs the potential benefit of leaving it around. 

I welcome viewpoints.

**Comment by dagewa** [comment]
I have wondered whether a simpler redesign of multiprocessing for integration of rotation sweeps would be more efficient than any of the current approaches. If we divided up the rotation sweep at a high level into N blocks (where N is the number of cores or maybe processor threads) and each of these went through the whole workflow of `dials.integrate` for 1/Nth of the rotation sweep as if `nproc=1` then might this not be much more efficient?

In practice, the N splits should overlap to avoid edge effects, probably by one whole block of reflection profile modelling. But essentially the global scan-varying model should allow separate regions of a rotation scan to be integrated independently. There would be a little bit of work to do afterwards to glue the results back together.

**Reply by graeme-winter** [reply]
This is how things work in `fp3` and there _is_ a measurable impact on the quality of the data. What we do at the moment is relatively sensible, just needs to be restructured a little (I think) to be more efficient.

**Reply by dagewa** [reply]
I don't see why there would be an impact on data quality, if each job overlapped by a full block of reflection profile modelling. Are we not then doing mathematically the same operations?

**Reply by graeme-winter** [reply]
Right, sorry, overlap. Yes, that makes sense.

However, that would take us to +/- a little what we already have right?

**Reply by dagewa** [reply]
Maybe... I actually mentioned because of the suspicion that we might be firing up many more subprocesses than previously expected (https://github.com/dials/dials/pull/1724#issuecomment-846962605). I don't really know how we currently do mp in integration ("threaded" or not)

**Reply by graeme-winter** [reply]
I think integration is probably a little more sane in this regard than the spot finding - we do seem to spin up a process per block which is probably not offensive. From what I can tell, for a substantial spot finding job where block size is 20, we spin up a process for each block of 20 images so this is potentially a substantial overhead. I also suspect that the difference between cores and threads is really hurting me in this case.
